rfcs/text/refactoring-date-component.md

```diff
@@ -0,0 +1,330 @@
+
+```js
+  inputProps = {
+    onFocus,
+    disabled = false,
+    readOnly = false,
+    autoFocus = false,
+    size = "medium",
+    labelHelp,
+    labelAlign,
+    labelInline,
+    labelSpacing,
+    labelWidth,
+    inputWidth,
+    fieldHelp,
+    error,
+    warning,
+    info,
+    placeholder,
+    required,
+    allowEmptyValue = false,
+    ...marginSpacingProps,
+  }
+```
+
+# Motivation
+
+The current component within `carbon-react` has a significant amount of technical debt and has a lot of logic specific to a project that was previously its main consumer. For example, the component supports being used as both as a controlled and uncontrolled component which again adds to the complexity. The current implementation is also tied to using `moment` to manage the parsing and formatting of date values. However, the maintainers of the [project](https://momentjs.com/docs/#/-project-status/) have previously labelled it as a legacy project and as such will no longer be releasing new features and it will only be maintained up to the point of addressing critical security concerns. Whilst this is not necessarily something that will have a significant impact on all our consumers, we should look to implement a `Date` component that does not rely on a legacy solution as there are several lightweight alternatives. Carbon currently exports an extensive set of [date utils](https://github.com/Sage/carbon/blob/master/src/utils/helpers/date/date.js) all of which use `moment` and support a large amount of date formats for all locales which adds to the complexity of the component.  
+
+As the number of consumers of the library has grown this has led to a lot of [issues](https://github.com/Sage/carbon/issues?q=is%3Aissue+is%3Aopen+date) being raised against the component. Some of the code was written over four years ago, and the addition of some newer features during this time have led to the code for the component being significantly bloated and therefore difficult to maintain.
+
+# Detailed design
+
+The date passed to the picker component can be updated through a `useEffect` hook that checks the current value in the input is both a valid format (based on the value recieved from the locale context) and parses to a valid Date. As such the component will be able to maintain a significantly reduced amount of internal state for the `selectedDate` and the `open` state for the picker component.
+
+```jsx
+  // src/components/date/date.component.js
+  import {
+    formatToString,
+    isValidFormat,
+    parseToDate,
+    isValidDate,
+    formatISOString
+  } from "./__internal__/date-utils"
+
+  const DateInput = ({ value, onChange, onBlur, onFocus, inputProps, ...rest }) => {
+    const locale = useContext(LocaleContext);
+    const { date } = locale;
+    const format = date.formats.javascript();
+    const formats = date.formats.input();
```

Comment:
Suggestion, not a blocker.
In carbon we would write:
```
const l = useContext(LocaleContext);
```
and just use like
```
l.date.formats.javascript()
```

Reply:
Yeah I just had it structured like this for readability here